### PR TITLE
Add caching for Path, BGPPath, IP and Prefix structures

### PR DESCRIPTION
**Pull request by taktv6**

### CODE REVIEW

File: protocols/bgp/packet/path_attributes.go [hunk 1]
```diff
@@ -37,6 +56,12 @@ func decodePathAttrs(buf *bytes.Buffer, tpal uint16, opt *DecodeOptions) (*PathA
 		}
 	}
 
+	if haveNextHop || haveOrigin || haveASPath {
+		if !haveNextHop || !haveOrigin || !haveASPath {
+			return nil, fmt.Errorf("A mandatory path attribute is missing")
```
Comment (hikhvar) [inline protocols/bgp/packet/path_attributes.go]: Lets ease debugging and print which mandatory path attribute is missing.

File: route/bgp_path_cache.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+func newBGPPathACache() *bgpPathACache {
+	return &bgpPathACache{
+		cache: make(map[BGPPathA]*BGPPathA, initialBGPPathACacheSize),
+	}
+}
+
+func (bgpc *bgpPathACache) get(p *BGPPathA) *BGPPathA {
+	bgpc.cacheMu.Lock()
+
+	if _, ok := bgpc.cache[*p]; ok {
```
Comment (hikhvar) [inline route/bgp_path_cache.go]: Same here: 
```
if x, ok :=... 
```

File: protocols/bgp/packet/path_attributes_test.go [hunk 3]
```diff
@@ -2032,7 +2050,7 @@ func TestSerialize(t *testing.T) {
 				22, 185, 65, 240, // 185.65.240.0/22
 			},
 		},
-		{
+		/*{
```
Comment (czerwonk) [inline protocols/bgp/packet/path_attributes_test.go]: intended?

File: routingtable/table_test.go [hunk 4]
```diff
@@ -8,13 +8,57 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+/*func TestAddPathFullTable(t *testing.T) {
```
Comment (czerwonk) [inline routingtable/table_test.go]: can this be removed?

File: util/servicewrapper/grpc.go [hunk 5]
```diff
@@ -76,7 +76,7 @@ func New(grpcPort uint16, h *http.Server, unaryInterceptors []grpc.UnaryServerIn
 	streamInterceptors = append(streamInterceptors,
 		grpc_prometheus.StreamServerInterceptor,
 		grpc_ctxtags.StreamServerInterceptor(),
-		grpc_recovery.StreamServerInterceptor(),
+		//grpc_recovery.StreamServerInterceptor(),
```
Comment (czerwonk) [inline util/servicewrapper/grpc.go]: duplicate, please remove

File: protocols/bgp/packet/decoder_test.go [hunk 6]
```diff
@@ -1247,54 +1159,6 @@ func TestDecodeUpdateMsg(t *testing.T) {
 				},
 			},
 		},
-		{
```
[the remark this answers is not in the document]
Comment (taktv6) [reply protocols/bgp/packet/decoder_test.go]: Testcase not valid anymore. Mandatory attribute missing.

File: protocols/bgp/packet/decoder_test.go [hunk 7]
```diff
@@ -565,74 +545,6 @@ func TestDecodeUpdateMsg(t *testing.T) {
 			},
 			wantFail: true,
 		},
-		{
```
[the remark this answers is not in the document]
Comment (taktv6) [reply protocols/bgp/packet/decoder_test.go]: Testcase not valid anymore. Mandatory attribute missing.

File: protocols/bgp/packet/path_attributes.go (hunk 1 above)
Comment (taktv6) [reply protocols/bgp/packet/path_attributes.go]: Done.

File: net/ip_cache.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+func newIPCache() *ipCache {
+	return &ipCache{
+		cache: make(map[IP]*IP, ipCacheInitialSize),
+	}
+}
+
+func (ipc *ipCache) get(addr *IP) *IP {
+	ipc.cacheMu.Lock()
+
+	if _, ok := ipc.cache[*addr]; ok {
```
[the remark this answers is not in the document]
Comment (taktv6) [reply net/ip_cache.go]: Done

File: net/prefix_cache.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+func newPfxCache() *pfxCache {
+	return &pfxCache{
+		cache: make(map[Prefix]*Prefix, prefixCacheInitialSize),
+	}
+}
+
+func (pfxc *pfxCache) get(pfx *Prefix) *Prefix {
+	pfxc.cacheMu.Lock()
+
+	if _, ok := pfxc.cache[*pfx]; ok {
```
[the remark this answers is not in the document]
Comment (taktv6) [reply net/prefix_cache.go]: Done.

File: route/bgp_path_cache.go (hunk 2 above)
Comment (taktv6) [reply route/bgp_path_cache.go]: Done.

File: util/servicewrapper/grpc.go (hunk 5 above)
Comment (taktv6) [reply util/servicewrapper/grpc.go]: Done.

File: routingtable/table_test.go (hunk 4 above)
Comment (taktv6) [reply routingtable/table_test.go]: Removed.

**Pull request merged**